apis/supervisor/idp/v1alpha1/types_githubidentityprovider.go.tmpl

```diff
@@ -0,0 +1,218 @@
+	// +kubebuilder:default=slug
+	// +kubebuilder:validation:Enum=name;slug
+	Groups GitHubGroupNamesSpec `json:"groups"`
+}
+
+// GitHubClientSpec contains information about the GitHub client that this identity provider will use
+// for web-based login flows.
+type GitHubClientSpec struct {
+	// SecretName contains the name of a namespace-local Secret object that provides the clientID and
+	// clientSecret for an GitHub App or GitHub OAuth2 client.
+	//
+	// This secret must be of type "secrets.pinniped.dev/github-client" with keys "clientID" and "clientSecret".
+	SecretName string `json:"secretName"`
+}
+
+// GitHubIdentityProviderSpec is the spec for configuring an GitHub identity provider.
+type GitHubIdentityProviderSpec struct {
+	// GitHubApi allows configuration for GitHub Enterprise Server
+	//
+	// +optional
+	GitHubApi GitHubApiConfig `json:"github_api,omitempty"`
+
+	// Claims allows customization of the username and groups claims
+	//
+	// +optional
+	Claims GitHubClaims `json:"claims,omitempty"`
+
+	// AllowedOrganizations, when specified, indicates that only users with membership in at least one of the listed
+	// GitHub organizations may log in. In addition, the group membership presented to Kubernetes
+	// will only include teams within the listed GitHub organizations. Additional login rules or group filtering
+	// can optionally be provided as policy expression on any FederationDomain that includes this IDP.
+	//
+	// The configured GitHub App or GitHub OAuth App must be allowed to see membership in the listed organizations,
+	// otherwise Pinniped will not be aware that the user belongs to the listed organization or any teams
+	// within that organization.
+	//
+	// If no organizations are listed, you must set allowAllUsersToLogin to "Allow".
+	//
+	// +optional
+	AllowedOrganizations []string `json:"allowedOrganizations"`
+
+	// AllowAllUsersToLogin must be set to "Allow" if allowedOrganizations is empty.
+	//
+	// This field only exists to ensure that Pinniped administrators are aware that an empty list of
+	// allowedOrganizations means all GitHub users are allowed to log in.
+	//
+	// +kubebuilder:default=Deny
+	// +kubebuilder:validation:Enum=Allow;Deny
+	AllowAllUsersToLogin GitHubAllowAllUsersToLoginSpec `json:"allowAllUsersToLogin"`
+
+	// Client identifies the secret with credentials for a GitHub App or GitHub OAuth2 App (a GitHub client).
+	Client GitHubClientSpec `json:"client"`
+}
+
+// GitHubIdentityProvider describes the configuration of an upstream GitHub identity provider.
+// This upstream provider can be configured with either a GitHub App or a GitHub OAuth2 App.
+//
+// Right now, only web-based logins are supported, for both the pinniped-cli client and dynamic clients.
+//
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:resource:categories=pinniped;pinniped-idp;pinniped-idps
+// +kubebuilder:printcolumn:name="Issuer",type=string,JSONPath=`.spec.issuer`
```

Comment:
There is no `spec.issuer` on this resource, right? Should we show spec.gitHubAPI.host` instead? Although when empty it might look weird?

Reply:
Maybe we should set a default value for `spec.gitHubAPI.host` so we can show it as a print column?

Reply:
I thought about adding that column and chose not to for the reason you suggest.

What if we did this? If the user does not provide `spec.gitHubAPI.host` then Pinniped will infer `github.com`. Why not make that part of the API of this CRD? It should be the default value and the CRD can make `github.com` the actual default value.

In that case, we could show `spec.gitHubAPI.host` as a column here.

Reply:
Good catch that `Issuer` is not actually here. That was a copy/paste mistake.

Reply:
We'll show the `host` column and default it to `github.com`.